Error on building from source on termux

`rustc` / `cargo`: 1.94.1 

`cargo build --no-default-features --release`

```
Compiling elle v1.0.0 (/data/data/com.termux/files/home/code/elle)
error[E0432]: unresolved import `platform`
   --> src/io/watch.rs:451:16
    |
451 | pub(crate) use platform::FsWatcher;
    |                ^^^^^^^^ use of unresolved module or unlinked crate `platform`
    |
    = help: if you wanted to use a crate named `platform`, use `cargo add platform` to add it to your `Cargo.toml`

error[E0425]: cannot find function `watch_read_blocking` in this scope
   --> src/io/threadpool.rs:415:45
    |
415 | ... fd } => watch_read_blocking(fd),
    |             ^^^^^^^^^^^^^^^^^^^ not found in this scope

warning: unused variable: `op_id`
   --> src/io/aio.rs:224:21
    |
224 |                 for op_id in ids_to_cancel {
    |                     ^^^^^ help: if this is intentional, prefix it with an underscore: `_op_id`
    |
    = note: `#[warn(unused_variables)]` (part of `#[warn(unused)]`) on by default

warning: unused variable: `timeout`
   --> src/io/aio.rs:610:50
    |
610 | ... &ConnectAddr, timeout: Option<Duration>) ->...
    |                   ^^^^^^^ help: if this is intentional, prefix it with an underscore: `_timeout`

warning: unused variable: `timeout`
   --> src/io/aio.rs:835:9
    |
835 |         timeout: Option<Duration>,
    |         ^^^^^^^ help: if this is intentional, prefix it with an underscore: `_timeout`

warning: unused variable: `network_pool`
   --> src/io/aio.rs:919:13
    |
919 |             ref mut network_pool,
    |             ^^^^^^^^^^^^^^^^^^^^ help: try ignoring the field: `network_pool: _`

Some errors have detailed explanations: E0425, E0432.
For more information about an error, try `rustc --explain E0425`.
warning: `elle` (lib) generated 4 warnings
error: could not compile `elle` (lib) due to 2 previous errors; 4 warnings emitted
```

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -150,6 +150,30 @@ jobs:
       - name: Run smoke tests
         run: make smoke
 
+  android:
+    name: Android Cross-Check
+    needs: changes
+    runs-on: ubuntu-24.04-arm
+    if: needs.changes.outputs.source == 'true'
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@stable
+        with:
+          targets: aarch64-linux-android
+      - uses: Swatinem/rust-cache@v2
+        with:
+          shared-key: android
+      - name: Cross-check Android target
+        run: cargo check --target aarch64-linux-android --no-default-features -p elle
+      - name: Install GNU parallel
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y parallel
+      - name: Build host binary (no JIT)
+        run: cargo build --release --no-default-features -p elle
+      - name: Run smoke tests (VM only)
+        run: make smoke-vm
+
   macos:
     name: macOS Build & Smoke
     needs: changes
@@ -170,7 +194,7 @@ jobs:
 
   all-checks:
     name: All Checks Passed
-    needs: [changes, qa, tests-combined, tests, mlir, aarch64, macos]
+    needs: [changes, qa, tests-combined, tests, mlir, aarch64, android, macos]
     runs-on: ubuntu-latest
     if: always()
     steps:
@@ -188,6 +212,9 @@ jobs:
         run: exit 1
       - name: Check aarch64
         if: needs.aarch64.result == 'failure'
+        run: exit 1
+      - name: Check Android
+        if: needs.android.result == 'failure'
         run: exit 1
       - name: Check macOS
         if: needs.macos.result == 'failure'

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -166,8 +166,8 @@ jobs:
       - name: Configure NDK toolchain
         run: |
           NDK=$(ls -d "$ANDROID_SDK_ROOT/ndk/"* 2>/dev/null | sort -V | tail -1)
-          HOST_TAG=$(uname -m)-linux
-          echo "$NDK/toolchains/llvm/prebuilt/linux-$HOST_TAG/bin" >> "$GITHUB_PATH"
+          PREBUILT=$(ls -d "$NDK/toolchains/llvm/prebuilt/"* | head -1)
+          echo "$PREBUILT/bin" >> "$GITHUB_PATH"
           echo "CC_aarch64_linux_android=aarch64-linux-android24-clang" >> "$GITHUB_ENV"
           echo "AR_aarch64_linux_android=llvm-ar" >> "$GITHUB_ENV"
       - name: Cross-check Android target

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -149,11 +149,15 @@ jobs:
         run: cargo build --release -p elle
       - name: Run smoke tests
         run: make smoke
+      - name: Build release binary (no JIT)
+        run: cargo build --release --no-default-features -p elle
+      - name: Run smoke tests (VM only, no JIT)
+        run: make smoke-vm
 
   android:
     name: Android Cross-Check
     needs: changes
-    runs-on: ubuntu-24.04-arm
+    runs-on: ubuntu-latest
     if: needs.changes.outputs.source == 'true'
     steps:
       - uses: actions/checkout@v4
@@ -163,26 +167,14 @@ jobs:
       - uses: Swatinem/rust-cache@v2
         with:
           shared-key: android
-      - uses: nttld/setup-ndk@v1
-        id: ndk
-        with:
-          ndk-version: r27d
       - name: Configure NDK toolchain
         run: |
-          PREBUILT=$(ls -d "${{ steps.ndk.outputs.ndk-path }}/toolchains/llvm/prebuilt/"* | head -1)
-          echo "$PREBUILT/bin" >> "$GITHUB_PATH"
+          NDK=$(ls -d "$ANDROID_SDK_ROOT/ndk/"* | sort -V | tail -1)
+          echo "$NDK/toolchains/llvm/prebuilt/linux-x86_64/bin" >> "$GITHUB_PATH"
           echo "CC_aarch64_linux_android=aarch64-linux-android24-clang" >> "$GITHUB_ENV"
           echo "AR_aarch64_linux_android=llvm-ar" >> "$GITHUB_ENV"
       - name: Cross-check Android target
         run: cargo check --target aarch64-linux-android --no-default-features -p elle
-      - name: Install GNU parallel
-        run: |
-          sudo apt-get update
-          sudo apt-get install -y parallel
-      - name: Build host binary (no JIT)
-        run: cargo build --release --no-default-features -p elle
-      - name: Run smoke tests (VM only)
-        run: make smoke-vm
 
   macos:
     name: macOS Build & Smoke

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -163,10 +163,13 @@ jobs:
       - uses: Swatinem/rust-cache@v2
         with:
           shared-key: android
+      - uses: nttld/setup-ndk@v1
+        id: ndk
+        with:
+          ndk-version: r27d
       - name: Configure NDK toolchain
         run: |
-          NDK=$(ls -d "$ANDROID_SDK_ROOT/ndk/"* 2>/dev/null | sort -V | tail -1)
-          PREBUILT=$(ls -d "$NDK/toolchains/llvm/prebuilt/"* | head -1)
+          PREBUILT=$(ls -d "${{ steps.ndk.outputs.ndk-path }}/toolchains/llvm/prebuilt/"* | head -1)
           echo "$PREBUILT/bin" >> "$GITHUB_PATH"
           echo "CC_aarch64_linux_android=aarch64-linux-android24-clang" >> "$GITHUB_ENV"
           echo "AR_aarch64_linux_android=llvm-ar" >> "$GITHUB_ENV"

--- a/.github/workflows/pr.yml
+++ b/.github/workflows/pr.yml
@@ -163,6 +163,13 @@ jobs:
       - uses: Swatinem/rust-cache@v2
         with:
           shared-key: android
+      - name: Configure NDK toolchain
+        run: |
+          NDK=$(ls -d "$ANDROID_SDK_ROOT/ndk/"* 2>/dev/null | sort -V | tail -1)
+          HOST_TAG=$(uname -m)-linux
+          echo "$NDK/toolchains/llvm/prebuilt/linux-$HOST_TAG/bin" >> "$GITHUB_PATH"
+          echo "CC_aarch64_linux_android=aarch64-linux-android24-clang" >> "$GITHUB_ENV"
+          echo "AR_aarch64_linux_android=llvm-ar" >> "$GITHUB_ENV"
       - name: Cross-check Android target
         run: cargo check --target aarch64-linux-android --no-default-features -p elle
       - name: Install GNU parallel

--- a/src/io/aio.rs
+++ b/src/io/aio.rs
@@ -221,11 +221,11 @@ impl AsyncBackend {
                     })
                     .collect();
 
-                for op_id in ids_to_cancel {
+                for _op_id in ids_to_cancel {
                     match inner.platform {
                         #[cfg(target_os = "linux")]
                         PlatformBackend::Uring(ref mut ring) => {
-                            let _ = crate::io::uring::submit_uring_cancel(ring, op_id);
+                            let _ = crate::io::uring::submit_uring_cancel(ring, _op_id);
                         }
                         PlatformBackend::ThreadPool(_) => {
                             // Thread pool: shutdown the fd to unblock any thread
@@ -607,6 +607,7 @@ impl AsyncBackend {
 
     /// Submit a Connect operation. Connect creates a new port, so
     /// request.port is Value::NIL — we handle it separately.
+    #[allow(unused_variables)]
     fn submit_connect(&self, addr: &ConnectAddr, timeout: Option<Duration>) -> Result<u64, String> {
         let mut inner = self.inner.borrow_mut();
         let id = inner.next_id;
@@ -782,6 +783,7 @@ impl AsyncBackend {
     }
 
     /// Submit a watch-next operation. Reads from the inotify fd.
+    #[allow(unused_variables)]
     fn submit_watch_next(&self, watcher_val: &Value) -> Result<u64, String> {
         use crate::io::watch::FsWatcher;
 
@@ -825,6 +827,7 @@ impl AsyncBackend {
 
     /// Submit a file open operation. Open creates a new port, so
     /// request.port is Value::NIL — we handle it before the port guard.
+    #[allow(unused_variables)]
     fn submit_open(
         &self,
         path: &str,
@@ -905,6 +908,7 @@ impl AsyncBackend {
         Ok(id)
     }
 
+    #[allow(unused_variables)]
     fn submit_task(&self, task_fn: &TaskFn) -> Result<u64, String> {
         let closure = task_fn
             .take()

--- a/src/io/threadpool.rs
+++ b/src/io/threadpool.rs
@@ -634,8 +634,8 @@ impl StdinThread {
     }
 }
 
-/// Blocking read on an inotify fd (Linux).
-#[cfg(target_os = "linux")]
+/// Blocking read on an inotify fd (Linux/Android).
+#[cfg(any(target_os = "linux", target_os = "android"))]
 fn watch_read_blocking(fd: RawFd) -> (i32, Vec<u8>) {
     let mut buf = vec![0u8; 4096];
     let ret = unsafe { libc::read(fd, buf.as_mut_ptr() as *mut libc::c_void, buf.len()) };

--- a/src/io/watch.rs
+++ b/src/io/watch.rs
@@ -35,7 +35,7 @@ impl WatchEventKind {
 
 // ─── Linux: inotify ────────────────────────────────────────────────────
 
-#[cfg(target_os = "linux")]
+#[cfg(any(target_os = "linux", target_os = "android"))]
 mod platform {
     use super::{WatchEvent, WatchEventKind};
     use std::cell::RefCell;
@@ -448,4 +448,5 @@ mod platform {
     }
 }
 
+#[cfg(any(target_os = "linux", target_os = "android", target_os = "macos"))]
 pub(crate) use platform::FsWatcher;

--- a/src/io/watch.rs
+++ b/src/io/watch.rs
@@ -135,7 +135,7 @@ mod platform {
                 .remove(&path)
                 .ok_or_else(|| format!("watch-remove: not watched: \"{}\"", path.display()))?;
             inner.wd_to_path.remove(&wd);
-            unsafe { libc::inotify_rm_watch(inner.fd, wd) };
+            unsafe { libc::inotify_rm_watch(inner.fd, wd as _) };
             Ok(())
         }
 

